ARROW-7836: [Rust] "allocate_aligned"/"reallocate" need to initialize memory to avoid UB

paddyhoran opened the pull request.

paddyhoran commented:
The CI failures are unrelated.  @Marwes @jturner314 are you happy with this solution?

jturner314 commented on rust/arrow/src/memory.rs:
```diff
@@ -38,11 +38,19 @@ pub fn free_aligned(p: *mut u8, size: usize) {
+        );
+        if new_size > old_size {
+            std::ptr::write_bytes(
+                new_ptr.offset(old_size as isize),
```
Fwiw, using the [`add`](https://doc.rust-lang.org/std/primitive.pointer.html#method.add) method is equivalent and a bit more concise:
```suggestion
                new_ptr.add(old_size),
```

jturner314 commented on rust/arrow/src/memory.rs:
```diff
@@ -38,11 +38,19 @@ pub fn free_aligned(p: *mut u8, size: usize) {
             new_size,
-        )
+        );
+        if new_size > old_size {
```
There should be an additional check here in case `realloc` [returns a null pointer (indicating an error)](https://doc.rust-lang.org/std/alloc/trait.GlobalAlloc.html#method.realloc), since otherwise the [`offset` will be out-of-bounds](https://doc.rust-lang.org/std/primitive.pointer.html#safety-1). (Also, `std::ptr::write_bytes` [requires the pointer to be non-null](https://doc.rust-lang.org/std/ptr/fn.write_bytes.html#safety).)
```suggestion
        if !new_ptr.is_null() && new_size > old_size {
```

jturner314 commented on rust/arrow/src/memory.rs:
```diff
@@ -38,11 +38,19 @@ pub fn free_aligned(p: *mut u8, size: usize) {
-        )
+        );
+        if new_size > old_size {
+            std::ptr::write_bytes(
```
Fwiw, using the [`.write_bytes()`](https://doc.rust-lang.org/std/primitive.pointer.html#method.write_bytes) method on the pointer would be a bit more concise.

paddyhoran commented:
I addressed all the comments.  @jturner314 thank you very much for all your time and help!

> I'd prioritize safety over potential performance gains.

Yea, definitely my thinking too.

paddyhoran commented:
@andygrove, @nevi-me @sunchao can I get at least one approval from another committer to get this merged if you have a chance?
